docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+> Clonable, or Serializable, or to define an equality relation on them. Some support for such functionality can be found
+> in third-party libraries, but the overall experience for the developer is too subpar.
+>
+> We thus settle on a different strategy: lifting chain-specific data into an enum over all possible chain types.
+
+Additionally, there are places where the core modules code (indirectly) depends on light client specific types. For
+e.g., in the code below, the `ClientEvents::try_from_tx()` function must be able to deserialize a light client specific
+header from the `UpdateClient` event.
+
+```rust
+pub fn from_tx_response_event(height: Height, event: &tendermint::abci::Event) -> Option<IbcEvent> {
+    if let Some(mut client_res) = ClientEvents::try_from_tx(event) {
+        client_res.set_height(height);
+        Some(client_res)
+    }
+    /* ... */
+}
+```
+
+As we look to add support for more light-clients, it would be much desirable for the core modules code to be
+light client agnostic allowing light client implementations to be hosted externally, and, maintained & audited
+independently.
+
+It would be possible to break the circular dependencies by extracting all core code (types, traits, etc.) that a
+light client implementation would possibly need into a separate crate, say `ibc-base`, and then have both the `ibc`
+modules crate and light client implementation crates depend on that `ibc-base` crate. This was implemented in PR
+[#2327](https://github.com/informalsystems/ibc-rs/pull/2327), but the solution seems to be fragile and unmaintainable.
+
+This ADR proposes to break the circular dependency problem by removing the `Any*` enums completely and using
+trait-objects in their place.
+
+## Decision
+
+### Object safety
+
+There are a total of 5 light client traits, each associated with an `Any*` enum that implements it ->
+
+* `AnyClient` implements `ClientDef`.
+* `AnyClientState` implements `ClientState`.
+* `AnyConsensusState` implements `ConsensusState`.
+* `AnyHeader` implements `Header`.
+* `AnyMisbehaviour` implements `Misbehaviour`.
+
+In order to replace the `Any*` enums with trait objects, these light client traits must be object safe. This
+essentially means that these traits cannot have a `Self: Sized` requirement and their methods cannot have type
+parameters & cannot use `Self`. These restrictions, their implications, and possible workarounds are documented below.
+
+#### Traits that require `Self: Sized` cannot be used as supertraits
+
+Common traits that the light client traits depend on that have the `Self: Sized` requirement
+are `core::convert::{Into<T>, From<T>, ...}`, `tendermint_proto::Protobuf`, `Clone`, etc.
+
+An easy workaround is to add these methods directly to the trait, for e.g. we add an `encode_vec()` method to
+the `Header` trait directly.
+
+```rust
+pub trait ClientState {
+    /// Encode to canonical binary representation
+    fn encode_vec(&self) -> Vec<u8>;
```

Comment:
This is an interesting idea and I thought this should be theoretically possible so I tried to experiment with it here -> 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0017151c945b0eebb412dbda9a06db20

The `Protobuf` trait requires the raw type to implement `Message + From<Self> + Default` and the domain type to implement `Sized + Clone + TryFrom<T>`. The raw type part is easy as prost already derives `Message` and `default` for us. For the domain types, we need `Clone` and `TryFrom` without the `Self: Sized` requirement, so we could write our own. And it all seems to work until we decide to call a `Protobuf` method on our trait object. 

This isn't allowed because methods such as `Protobuf::encode_vec()` must be able to convert the domain type to a raw type and at that point, the domain type's size must be known. This is the problem that the code in the playground runs into. I can't think of a good solution to this problem now but happy to revisit it if we find a decent workaround.

Reply:
Added this to the ADR -> https://github.com/informalsystems/ibc-rs/pull/2356/commits/0b085570e24119a7f921f178a156ce4dfaa97507